crates/stages/stages/src/stages/s3/mod.rs

```diff
@@ -0,0 +1,289 @@
+pub use downloader::{fetch, Metadata};
+use downloader::{DownloaderError, S3DownloaderResponse};
+
+mod filelist;
+use filelist::DOWNLOAD_FILE_LIST;
+
+use reth_db::transaction::DbTxMut;
+use reth_primitives::StaticFileSegment;
+use reth_provider::{
+    DBProvider, StageCheckpointReader, StageCheckpointWriter, StaticFileProviderFactory,
+};
+use reth_stages_api::{
+    ExecInput, ExecOutput, Stage, StageCheckpoint, StageError, StageId, UnwindInput, UnwindOutput,
+};
+use std::{
+    path::PathBuf,
+    task::{ready, Context, Poll},
+};
+use tokio::sync::mpsc::{unbounded_channel, UnboundedReceiver};
+
+/// S3 `StageId`
+const S3_STAGE_ID: StageId = StageId::Other("S3");
+
+/// The S3 stage.
+#[derive(Default, Debug)]
+#[non_exhaustive]
+pub struct S3Stage {
+    /// Static file directory.
+    static_file_directory: PathBuf,
+    /// Remote server URL.
+    url: String,
+    /// Maximum number of connections per download.
+    max_concurrent_requests: u64,
+    /// Channel to receive the downloaded ranges from the fetch task.
+    fetch_rx: Option<UnboundedReceiver<Result<S3DownloaderResponse, DownloaderError>>>,
+}
+
+impl<Provider> Stage<Provider> for S3Stage
+where
+    Provider: DBProvider<Tx: DbTxMut>
+        + StaticFileProviderFactory
+        + StageCheckpointReader
+        + StageCheckpointWriter,
+{
+    fn id(&self) -> StageId {
+        S3_STAGE_ID
+    }
+
+    fn poll_execute_ready(
+        &mut self,
+        cx: &mut Context<'_>,
+        input: ExecInput,
+    ) -> Poll<Result<(), StageError>> {
+        // We are currently fetching and may have downloaded ranges that we can process.
+        if let Some(rx) = &mut self.fetch_rx {
+            // Whether we have downloaded all the required files.
+            let mut is_done = false;
+
+            let response = match ready!(rx.poll_recv(cx)) {
+                Some(Ok(response)) => {
+                    is_done = response.is_done();
+                    Ok(())
+                }
+                Some(Err(_)) => todo!(), // TODO: DownloaderError -> StageError
+                None => Err(StageError::ChannelClosed),
+            };
+
+            if is_done {
+                self.fetch_rx = None;
+            }
+
+            return Poll::Ready(response)
+        }
+
+        // Spawns the downloader task if there are any missing files
+        if let Some(fetch_rx) = self.maybe_spawn_fetch(input) {
+            self.fetch_rx = Some(fetch_rx);
+            return Poll::Pending
+        }
+
+        Poll::Ready(Ok(()))
+    }
+
+    fn execute(&mut self, provider: &Provider, input: ExecInput) -> Result<ExecOutput, StageError>
+    where
+        Provider: DBProvider<Tx: DbTxMut>
+            + StaticFileProviderFactory
+            + StageCheckpointReader
+            + StageCheckpointWriter,
+    {
+        // Re-initializes the provider to detect the new additions
+        provider.static_file_provider().initialize_index()?;
+
+        // TODO logic for appending tx_block
```

Comment:
is this still a todo or is init index all we need?

should this reinit per invocation or only once after we previously downloaded the file?